### PR TITLE
refactor(data): IBatchProvider over InputHelper (fixes #282)

### LINKED ISSUE #282: Dataset and DataLoader Abstractions for Images, Audio, and Video
Background
- Robust IO and preprocessing across modalities is essential for training.

Problem
- No unified dataset/dataloader abstractions for images/audio/video.

User Story
- As a practitioner, I want simple, performant dataset and dataloader APIs for common modalities with transforms.

Acceptance Criteria
- Datasets:
  - ImageFolder w/ transforms; AudioFolder (resample, mel); VideoFolder (frame decode, sampling strategies).
- Dataloaders:
  - Shuffling, bucketing for sequence lengths, prefetch, worker threads.
- Transforms:
  - Augmentations per modality; normalization; padding.
- Tests:
  - IO, determinism, speed sanity checks.
- Docs:
  - Recipes for common datasets (CIFAR-10, LJSpeech, WebVid subsets).

Dependencies
- None; supports many issues above.

Effort
- 1–2 weeks.

### PR DESCRIPTION
**Pull request by ooples**
Replaces custom loader/dataset with a tiny IBatchProvider<T,TInput> over existing InputHelper. Keeps everything generic and aligned with project architecture. Tests updated. Ready for review.

### CODE REVIEW

**Comment by ooples** [comment]
Update: Refactored DataLoader/Dataset to use Vector<T> and kept one-type-per-file organization. Confirmed tests on net462/net8.0. Will further integrate with PredictionModelBuilder patterns if consumption requires builder hooks.

**Comment by ooples** [comment]
## Closing this PR

After a comprehensive analysis of the codebase (40K+ lines across 300+ files), this PR has been determined to add unnecessary abstraction without solving a real problem.

### Why This PR Doesn't Add Value

**1. Pure Delegation - No New Functionality**
`IBatchProvider` literally just wraps `InputHelper` with zero new functionality:
```csharp
public sealed class DefaultBatchProvider<T, TInput> : IBatchProvider<T, TInput>
{
    public int GetBatchSize(TInput input) => InputHelper<T, TInput>.GetBatchSize(input);
    public TInput GetBatch(TInput input, int[] indices) => InputHelper<T, TInput>.GetBatch(input, indices);
}
```

**2. Solves a Non-Existent Problem**
- `InputHelper` is used 51+ times across the codebase and works perfectly
- All optimizers, cross-validators, and feature selectors use it successfully
- No one has complained about `InputHelper` - it's not broken

**3. Architectural Mismatch**
- Models have concrete signatures: `void Train(Matrix<T> x, Vector<T> y)`
- They don't use generic `TInput`/`TOutput`, so this interface serves no purpose

**4. No Extensibility Benefit**
- There's no scenario where someone would implement `IBatchProvider` differently
- All batching logic is already handled by `InputHelper.GetBatch`

### The Real Issue (If Any)

The actual problem is **repetitive batching code in optimizers** (40+ instances of manual shuffle + batch loops), not the lack of an interface.

### Better Solution: Issue #443

Created **#443** which proposes a `DataLoader` utility that:
- ✅ Eliminates 40+ instances of duplicate batching code
- ✅ Provides `IEnumerable<(TInput, TOutput)>` for clean iteration
- ✅ Handles shuffling, drop_last, seeding consistently
- ✅ Works with existing `InputHelper` infrastructure
- ✅ Enables advanced features (stratified batching, weighted sampling)

**Example:**
```csharp
// Before (repeated 40+ times):
for (int epoch = 0; epoch < maxEpochs; epoch++)
{
    var shuffled = ShuffleIndices(indices, random);
    for (int i = 0; i < numBatches; i++)
    {
        var batchIndices = shuffled.Skip(i * batchSize).Take(batchSize).ToArray();
        var xBatch = InputHelper<T, TInput>.GetBatch(xTrain, batchIndices);
        var yBatch = InputHelper<T, TOutput>.GetBatch(yTrain, batchIndices);
        // ... training
    }
}

// After (with DataLoader):
for (int epoch = 0; epoch < maxEpochs; epoch++)
{
    foreach (var (xBatch, yBatch) in (xTrain, yTrain).CreateBatches<T>(batchSize, shuffle: true))
    {
        // ... training
    }
}
```

### Recommendation

**Close this PR** and implement #443 instead, which addresses the real problem (code duplication) rather than adding an unnecessary abstraction layer.

---

**Reference:** Deep codebase analysis available upon request. This decision is based on reviewing:
- 51 usages of `InputHelper` across optimizers, cross-validators, feature selectors
- All model training signatures (concrete `Matrix<T>`/`Vector<T>`, not generic `TInput`/`TOutput`)
- Actual pain points in optimizer implementations

**See:** #443 for the better alternative that actually solves the batching code duplication problem.
